### DISCUSSION TITLE
How can I access "TasksToExecute" in Cake Frosting?

### DISCUSSION THREAD

**Discussion by boinst** [cake-build/cake#3773] [Q&A]
My quest started [here](https://github.com/cake-build/cake/issues/1772).

I am using Cake Frosting.

I would like to be able to access a list of tasks that are going to be executed.

My use case is, I have two tasks, but only one should be executed. Task A does [XYZ], and task B does only [Y]. So if task A is going to be executed, I don't need to run B. What I am really looking for is a way to skip B if A is going to be executed. The property "TasksToExecute" seems to be a suitable candidate, but I can't figure out how to access it in Cake Frosting.

Thanks in advance.

**Comment by augustoproiete** [comment]
Not exactly a direct answer to your question on accessing `TasksToExecute`, but if `XYZ` are not individual Frosting Tasks (and just methods), then an easy way would be to use a property in your build context (`FrostingContext`) to track whether `Y` already ran or not...

```csharp
public class BuildContext : FrostingContext
{
    public bool YAlreadyExecuted { get; set; }

    public BuildContext(ICakeContext context)
        : base(context)
    {
    }
}
```

And then share the context across `X`, `Y`, and `Z`:

```csharp
[TaskName("TaskA")]
public sealed class TaskA : FrostingTask<BuildContext>
{
    public override void Run(BuildContext context)
    {
        Steps.X(context);
        Steps.Y(context);
        Steps.Z(context);
    }
}

[TaskName("TaskB")]
public sealed class TaskB : FrostingTask<BuildContext>
{
    public override void Run(BuildContext context)
    {
        // You could check context.YAlreadyExecuted here if you wanted
        Steps.Y(context);
    }
}
```

Up to you whether you do the check in `TaskB` or directly with `Y` to ensure it runs only once...

```csharp
public static class Steps
{
    public static void X(BuildContext context)
    {
        context.Log.Information("X");
    }

    public static void Y(BuildContext context)
    {
        if (context.YAlreadyExecuted)
        {
            return;
        }

        context.Log.Information("Y");
        context.YAlreadyExecuted = true;
    }

    public static void Z(BuildContext context)
    {
        context.Log.Information("Z");
    }
}
```

**Reply by boinst** [reply]
Thank you @augustoproiete . That's pretty close to what I'm doing now. Where it falls down is if `TaskB` (step `Y`) executes first, when we get to executing `TaskA`, this requires that `X`, `Y` and `Z` are executed _in sequence_. In my case, `X` and `Z` start and terminate a background process (SonarQube) that monitors what `Y` (an MSBuild execution) is doing. To get around that, I end up needing a command-line argument something like `--execute-x-and-z`. If I had access to `TasksToExecute`, I could eliminate the command-line argument.
